lib/md054.js

```diff
@@ -0,0 +1,134 @@
+// @ts-check
+
+"use strict";
+
+const { addErrorContext } = require("../helpers");
+const { filterByTypes, filterByPredicate, getTokenTextByType } =
+  require("../helpers/micromark.cjs");
+
+const isInlineLink = ({ children }) => children.some(
+  ({ type }) => type === "resource"
+);
+
+const isAutolink = ({ type }) => type === "autolink";
+
+const getNestedTokenTextByType = (tokens, type) => getTokenTextByType(
+  filterByTypes(tokens, [ type ]),
+  type
+);
+
+const escapeParentheses = (unescaped) => unescaped
+  .replaceAll("(", "\\(")
+  .replaceAll(")", "\\)");
+const escapeSquares = (unescaped) => unescaped
+  .replaceAll("[", "\\[")
+  .replaceAll("]", "\\]");
+const escapeAngles = (unescaped) => unescaped
+  .replaceAll("<", "\\<")
+  .replaceAll(">", "\\>");
+
+const unescapeParentheses = (escaped) => escaped
+  .replaceAll("\\(", "(")
+  .replaceAll("\\)", ")");
+const unescapeAngles = (escaped) => escaped
+  .replaceAll("\\<", "<")
+  .replaceAll("\\>", ">");
+
+const referenceLinkDestination = (link, tokens) => {
+  const reference = getNestedTokenTextByType([ link ], "reference");
+  const id = reference && reference !== "[]" ?
+    reference.replace(/^\[/, "").replace(/\]$/, "") :
+    getNestedTokenTextByType([ link ], "labelText");
+
+  const definition = filterByPredicate(
+    filterByTypes(tokens, [ "definition" ]),
+    (d) => getNestedTokenTextByType([ d ], "definitionLabelString") === id
+  );
+  return getNestedTokenTextByType(definition, "definitionDestination");
+};
+
+const inlineLinkDestination = (link) => {
+  const text = getNestedTokenTextByType([ link ], "resourceDestination");
+  return text && unescapeParentheses(text);
+};
+
+const autolinkDestination = (link) => {
+  const text = getNestedTokenTextByType([ link ], "autolinkProtocol");
+  return text && unescapeAngles(text);
```

Comment:
I'm not SURE this is the right place? Characters can be escaped in various ways. I'm not sure we should assume only angles are escaped in autolink, etc.. I think I'd rather escape what's necessary on write instead of on read?